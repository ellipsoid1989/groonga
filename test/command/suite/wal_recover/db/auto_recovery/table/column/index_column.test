# _database_close/_database_reopen can't use with HTTP because requests via HTTP are always
# processed by workers. Workers never use primary WAL role.
#@require-interface stdio

#@require-feature message_pack

#$GROONGA_ALLOW_DATABASE_REOPEN=yes
#$GRN_WAL_ROLE=primary

table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

load --table Data
[
{"_key": "a",      "content": "a"},
{"_key": "abc",    "content": "a and b and c"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]

delete Data --key abc

table_create Terms TABLE_PAT_KEY ShortText \
  --default_tokenizer TokenNgram \
  --normalizer NormalizerNFKC130
column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content

dump

_database_close

#@copy-path #{db_path}.0000100 #{db_path}.0000100.wal
#@add-important-log-levels notice
_database_reopen
#@remove-important-log-levels notice

dump

index_column_diff Terms data_content

select Data \
  --match_columns 'content * 2' \
  --query e \
  --output_columns _id,_key,content,_score
