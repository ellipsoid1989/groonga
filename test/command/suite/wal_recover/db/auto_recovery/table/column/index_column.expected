table_create Data TABLE_PAT_KEY ShortText
[[0,0.0,0.0],true]
column_create Data content COLUMN_SCALAR ShortText
[[0,0.0,0.0],true]
load --table Data
[
{"_key": "a",      "content": "a"},
{"_key": "abc",    "content": "a and b and c"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]
[[0,0.0,0.0],3]
delete Data --key abc
[[0,0.0,0.0],true]
table_create Terms TABLE_PAT_KEY ShortText   --default_tokenizer TokenNgram   --normalizer NormalizerNFKC130
[[0,0.0,0.0],true]
column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
[[0,0.0,0.0],true]
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

load --table Data
[
["_key","content"],
["a","a"],
["abcdef","a and b and c and d and e and f"]
]

column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
_database_close
[[0,0.0,0.0],true]
_database_reopen
[[0,0.0,0.0],true]
#|e| [reader][read-entry][pat][recover][Data] must be map: <positive-integer>(2): path:<db/db.0000100>
#|n| [db][wal][recover] rebuild broken table: <Data>(256)
#|n| DDL:260:table_create #recovering#Data
#|n| spec:260:update:#recovering#Data:49(table:pat_key):0
#|n| [db][wal][recover] rebuild broken column: <Data.content>(257)
#|n| DDL:261:column_create #recovering#Data content
#|n| spec:261:update:#recovering#Data.content:65(column:var_size):14(ShortText)
#|n| [db][wal][recover] rebuild broken column: <Terms.data_content>(259)
#|n| DDL:262:column_create Terms #recovering#data_content
#|n| spec:262:update:Terms.#recovering#data_content:72(column:index):260(#recovering#Data)
#|n| DDL:262:set_source Terms.#recovering#data_content #recovering#Data.content
#|n| spec:261:update:#recovering#Data.content:65(column:var_size):14(ShortText)
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| DDL:2147483649:set_normalizers NormalizerNFKC130
#|n| spec:262:update:Terms.#recovering#data_content:72(column:index):260(#recovering#Data)
#|n| DDL:259:obj_remove Terms.data_content
#|n| spec:257:update:Data.content:65(column:var_size):14(ShortText)
#|n| spec:259:remove:Terms.data_content:72(column:index)
#|n| [db][wal][recover] succeeded to rebuild broken column: <Terms.data_content>(262)
#|n| DDL:256:obj_remove Data
#|n| DDL:257:obj_remove Data.content
#|n| spec:257:remove:Data.content:65(column:var_size)
#|n| spec:256:remove:Data:49(table:pat_key)
#|n| [db][wal][recover] succeeded to rebuild broken table: <Data>(260)
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

load --table Data
[
["_key","content"],
["a","a"],
["abcdef","a and b and c and d and e and f"]
]

column_create Terms data_content COLUMN_INDEX|WITH_POSITION Data content
index_column_diff Terms data_content
[[0,0.0,0.0],[]]
select Data   --match_columns 'content * 2'   --query e   --output_columns _id,_key,content,_score
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "content",
          "ShortText"
        ],
        [
          "_score",
          "Int32"
        ]
      ],
      [
        2,
        "abcdef",
        "a and b and c and d and e and f",
        2
      ]
    ]
  ]
]
